### PR TITLE
Improve deserialization performance

### PR DESCRIPTION
**Pull request by jrmuizel**
I haven't yet tested how much this improves deserialization performance but it should be noticeable.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/1834)

### CODE REVIEW

File: webrender_api/src/display_list.rs [hunk 1]
```diff
@@ -179,6 +179,45 @@ fn skip_slice<T: for<'de> Deserialize<'de>>(
     (range, count)
 }
 
+struct UnsafeReader<'a: 'b, 'b> {
```
[the remark this answers is not in the document]
Comment (Gankro) [reply webrender_api/src/display_list.rs]: Also why it's desirable.

**Comment by jrmuizel** [comment]
With this version and an updated bincode (https://github.com/TyOverby/bincode/pull/207) I go from 140fps to 200fps on my gmail yaml

File: webrender_api/src/display_list.rs [hunk 2]
```diff
@@ -525,6 +527,73 @@ fn serialize_fast<T: Serialize>(vec: &mut Vec<u8>, e: &T) {
     debug_assert!(((w.0 as usize) - (vec.as_ptr() as usize)) == vec.len());
 }
 
+// This uses a (start, end) representation instead of (start, len) so that
+// only need to update a single field as we read through it. This gives
```
Comment (Gankro) [inline webrender_api/src/display_list.rs]: "gives" at end of line is junk

File: webrender_api/src/display_list.rs [hunk 3]
```diff
@@ -525,6 +527,73 @@ fn serialize_fast<T: Serialize>(vec: &mut Vec<u8>, e: &T) {
...
+    // Second, we only need to adjust 'start' after reading and it's only adjusted by a
+    // constant. This allows LLVM to avoid adjusting the length field after ever read
+    // and lets it be aggregated into a single adjustment.
+    #[inline(always)]
+    fn read_internal(&mut self, buf: &mut [u8]) {
+        // this is safe because we panic if start + buf.len() > end
+        unsafe {
+            if self.start.offset(buf.len() as isize) > self.end {
+                panic!();
+            }
```
Comment (Gankro) [inline webrender_api/src/display_list.rs]: This if should just be 

`assert!(self.start.offset(buf.len() as isize) <= self.end, "UnsafeReader: wrote past end of target");`

File: webrender_api/src/display_list.rs [hunk 4]
```diff
@@ -525,6 +527,73 @@ fn serialize_fast<T: Serialize>(vec: &mut Vec<u8>, e: &T) {
     debug_assert!(((w.0 as usize) - (vec.as_ptr() as usize)) == vec.len());
 }
 
+// This uses a (start, end) representation instead of (start, len) so that
+// only need to update a single field as we read through it. This gives
+// makes it easier for llvm to understand what's going on. (https://github.com/rust-lang/rust/issues/45068)
+// we update the slice only once we're done reading
+struct UnsafeReader<'a: 'b, 'b> {
```
Comment (Gankro) [inline webrender_api/src/display_list.rs]: This is actually totally save to use, so maybe just ExactReader?

**Comment by glennw** [comment]
Wow, that's a nice FPS improvement!

**Comment by Gankro** [comment]
r=me with requested changes

**Comment by hherman1** [comment]
Hi! Rust noob here,

Would it make sense to make this available in the rust stl if it’s so much faster than the built in?

**Comment by Gankro** [comment]
Probably not, for a few reasons. It's specific to how we're using serde (which isn't part of std), and is also unsound if you're not using serde exactly like us. Basically we're relying on the fact that all of our serialization code is machine-generated, so we can't ever get expected and actual size wrong.

**Pull request merged**